GPT/gpt-function-calling.py

```diff
@@ -0,0 +1,185 @@
+                        "properties": {
+                            "str": {
+                                "type": "string",
+                                "description": "The command to search for",
+                            }
+                        },
+                        "required": ["str"],
+                    },
+                },
+            }
+        ],
+        "max_tokens": 2024,
+        "temperature": 0.6,
+        "n": 1,
+        "stop": None,
+        "model": settings.get("user.openai_model"),
+    }
+
+
+    response = requests.post(url, headers=headers, data=json.dumps(data))
+
+    if response.status_code == 200:
+        notify("GPT Task Completed")
+        print(response.json())
+
+        process_function_calls(insert_response, response)
+
+        content = (response.json()["choices"][0]["message"]["content"] or "").strip()
+        if len(content) != 0:
+            actions.user.display_response(content)
+
+    else:
+        notify("GPT Failure: Check API Key, Model, or Prompt")
+        print(response.json())
+
+def process_function_calls(insert_response, response):
+    try:
+        tool_calls = response.json()["choices"][0]["message"]["tool_calls"]
+        while tool_calls:
+            tool = tool_calls.pop()
+            first_argument = tool['function']['arguments']
+            try:
+                first_argument = json.loads(first_argument)['str']
+            except Exception as e:
+                notify(f"Argument Jason was malformed: {e}")
+            match tool['function']['name']:
+                case 'insert':
+                    insert_response(first_argument)
+                case 'display':
+                    actions.user.display_response(first_argument)
+                case 'notify':
+                    actions.user.notify_user(first_argument)
+                case 'search_for_command':
+                    actions.user.search_for_command(first_argument)
+    except Exception as e:
+        notify(f"No tool_calls found in response from LLM: {e}")
+
+@mod.action_class
+class UserActions:
+    def gpt_go(utterance: str, selected_text: str) -> str:
+        """Answer an arbitrary question"""
+
+        return gpt_function_query(utterance, selected_text, actions.user.paste)
+
+    def gpt_go_cursorless(utterance: str, text_to_process: str, cursorless_destination: any):
+        """Apply a cursorless prompt"""
+        def insert_to_destination(result: str):
+            actions.user.cursorless_insert(cursorless_destination, result)
+        return gpt_function_query(utterance, text_to_process, insert_to_destination)
+
+    def notify_user(response: str):
+        """Send a notification to the desktop"""
+        actions.app.notify(response)
+
+    def search_for_command(response: str):
+        """Search VSCode for command"""
+        actions.user.command_palette()
+        actions.user.paste(response)
+
+    def display_response(response: str):
```

Comment:
would be cool to let gpt just output arbitrary html to be rendered. need to think about security implications and def out of scope for this PR